### PR TITLE
C22 Phoenix - Astry Brana - React Chatlog

### PR DESCRIPTION
**Pull request by SpectreKitty**

### CODE REVIEW

File: package-lock.json
Comment (anselrognlie) [inline package-lock.json:1]: Huh. I wonder what caused this change, since it doesn't look like there were any change sto your packages.json file itself. 🤔

File: src/components/ChatEntry.jsx [hunk 1]
```diff
@@ -1,20 +1,40 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp';
+
+const ChatEntry = ({ id, sender, body, timeStamp, liked, onLikeChat }) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:5]: 👍 I like using destructured props to make it more visibly clear in the function definition itself what props we're expecting to receive. We do need to remember that these are all passed in as a single object (the one we usually call `props`) and it can cause problems if we forget to include the destructuring syntax (it's easy to forget and list the props as multiple separate params) I really prefer the glanceability.

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,40 @@
...
-    <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={`chat-entry ${chatEntryClass()}`}>
+      <h2 className="entry-name">{sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time"><TimeStamp time={timeStamp}></TimeStamp></p>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:24]: Nice use of the supplied `TimeStamp`. All we need to do is pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

Note that the `TimeStamp` component doesn't receive children components (it resembles "void" HTML tags such as `input` or `img` rather than tags like `p` or `a` that can have children). For such components, prefer to write them using "self-closing" style (note the `/>` to complete the tag rather than a regular `>`).
```html
<TimeStamp time={timeStamp} />
```

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,40 @@
...
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  liked: PropTypes.bool.isRequired,
+  onLikeChat: PropTypes.func,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:37]: 👍 The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`, as you did. 

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatLog.jsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+          sender={chatEntry.sender}
+          body={chatEntry.body}
+          timeStamp={chatEntry.timeStamp}
+          liked={chatEntry.liked}
+          onLikeChat={onLikeChat}>
+        </ChatEntry>
+      </li>
+    );
+  });
+  return (
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:20]: Consider adding a blank line before the `return` to help visually separate the pre-calculation from the JSX result part.

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import './ChatLog.css';
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = ({ entries, onLikeChat }) => {
+  const chatEntryComponents = entries.map((chatEntry) => {
+    return (
+      <li key={chatEntry.id}>
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:8]: Nice job wrapping `ChatEnty` in a `li` tag. Since you're using `ul` to contain your entries, the direct children should be `li` tags once all the components have been expanded to HTML. Since `ChatEntry` is using `div` for its top-level tag, wrapping it while mapping the data provides that structure.

Good job recognizing that the `key` value then goes on the `li` item rather than our `ChatEntry`. Any tag that's rendered into a list, whether a custom component of our own, or a basic component provided by React, should be given a `key` value.

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import './ChatLog.css';
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = ({ entries, onLikeChat }) => {
+  const chatEntryComponents = entries.map((chatEntry) => {
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:6]: Nice use of `map` to convert from the message data into `ChatEntry` components. We can perform this mapping storing the result into a variable we use in the JSX result as you did here (components are functions, so we can run JS code as usual before we reach the `return`, and even sometimes have multiple `return` statements with different JSX), we could make a helper function that we call as part of the `return`, or this expression itself could be part of the `return` JSX, which I often like since it helps me see the overall structure of the component, though can make debugging a little more tricky. But any of those approaches will work fine.

File: src/App.jsx [hunk 7]
```diff
@@ -1,14 +1,42 @@
 import './App.css';
+import { useState } from 'react';
+import messagesData from './data/messages.json';
+import ChatLog from './components/ChatLog';
 
 const App = () => {
+  const [chatData, setChatData] = useState(messagesData);
+
+  const handleLikeChat = (id) => {
```
Comment (anselrognlie) [inline src/App.jsx:9]: 👍 Since our state is defined here, we also need to define our mutating function here so that it can "see" the setter function. All we need to receive is the `id` of the message to toggle, which allows us to locate the message to update, and calculate the next state value. We can pass this all the way down to our `ChatEntry` which handles the click event, passing us the `id` of the message that was clicked.

File: src/App.jsx [hunk 8]
```diff
@@ -1,14 +1,42 @@
...
+    setChatData(chatData => chatData.map(chat => {
```
Comment (anselrognlie) [inline src/App.jsx:10]: Nice use of the callback setter style. In this application, it doesn't really matter whether we use the callback style or the value style, but it's good practice to get in the habit of using the callback style.

Comment (anselrognlie) [reply src/App.jsx:10]: Nice use of map here to both handle making a new list so that React sees the message data has changed, and make new data for the clicked message with its like status toggled.

File: src/App.jsx [hunk 9]
```diff
@@ -1,14 +1,42 @@
...
+      if (chat.id == id) {
+        return {...chat, liked: !chat.liked};
```
Comment (anselrognlie) [inline src/App.jsx:12]: We showed this approach in class, but technically, we're mixing a few responsibilities here. rather than this function needing to know _how_ to change the liked status itself, we could move this update logic to a helper function. This would better mirror how we eventually update records when there's an API call involved.

In this project, our messages are very simple objects, but if we had more involved operations, it could be worthwhile to create an actual class with methods to work with them, or at least have a set of dedicated helper functions to centralize any such mutation logic.

File: src/App.jsx [hunk 10]
```diff
@@ -1,14 +1,42 @@
...
+    let totalLikes = 0;
+    for (const chat of chatData) {
+      if (chat.liked === true){
+        totalLikes += 1;
+      }
+    }
+    return totalLikes;
+  };
+
+  const totalLikes = calculateTotalLikeCount(chatData);
```
Comment (anselrognlie) [inline src/App.jsx:29]: Nice job determining the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

Explicitly totalling the count is perfectly fine, but many JS programmers would use `reduce` to achieve this.
```js
  const calculateTotalLikeCount = (chatData) => {
    return chatData.reduce((acc, chat) => {
      return chat.liked ? acc + 1 : acc;
    }, 0);
  };
```
The first few times we work with `reduce`, it can be challenging to understand, but it's a tool that gets used commonly enough that it's worth practicing.

File: src/App.jsx [hunk 11]
```diff
@@ -1,14 +1,42 @@
...
+      } else {
+        return chat;
+      }
+    }));
+  };
+
+  const calculateTotalLikeCount = (chatData) => {
+    let totalLikes = 0;
+    for (const chat of chatData) {
+      if (chat.liked === true){
```
Comment (anselrognlie) [inline src/App.jsx:22]: Since `liked` is already a boolean, just use its value directly without the comparison
```js
      if (chat.liked) {
```

File: src/components/ChatEntry.jsx [hunk 12]
```diff
@@ -1,20 +1,40 @@
...
+  };
+
+  const chatEntryClass = () => {
+    if (sender === 'Vladimir') {
+      return 'local';
+    }
+    else {
+      return 'remote';
+    }
+  };
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:17]: Nice logic to decide whether to treat a message as local or remote. How could we generalize this so that it didn't need to look explicitly for Vladimir? This project only passes in a single conversation, but ideally, our components should work in other situations.

In the general case, does the `ChatEntry` itself have enough information as it is to "know" which messages are local and which are remote?

File: src/App.jsx [hunk 13]
```diff
@@ -1,14 +1,42 @@
...
+    return totalLikes;
+  };
+
+  const totalLikes = calculateTotalLikeCount(chatData);
+
   return (
     <div id="App">
       <header>
-        <h1>Application title</h1>
+        <h1>Total Number of Likes: {totalLikes} ❤️s </h1>
```
Comment (anselrognlie) [inline src/App.jsx:34]: 👀 Notice in the example screenshots that in addition to the like counts, there was a header listing the conversation participants that we wanted, even if it weren't calculated dynamically.

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
